### PR TITLE
feat: use blockly indexes

### PR DESCRIPTION
**Pull request by alanshaw**
This PR adds support for "block link indexes" as a fallback, if content cannot be served by DUDEWHERE and SATNAV.

It binds a new bucket `BLOCKLY` (BLOCK+LYnks) and adds a new index implementation that uses the files stored in it. 

Blockly is a collection of [multi-indexes](https://github.com/web3-storage/specs/blob/73c386b999cf30fb648987ff9dce0516c1d91137/CARv2%20MultiIndex.md) for individual blocks. Each index contains the CID of a CAR file the block may be found in and the block's byte offset. The index also contains the _same detail for any links_ the block has.

### CODE REVIEW

File: src/lib/car-index.js [hunk 1]
```diff
@@ -147,3 +151,71 @@ export class StreamingCarIndex {
...
+  constructor (indexBucket) {
+    this.#bucket = indexBucket
+    /** @type {Map<RawCIDString, IndexEntry>} */
+    this.#cache = new Map()
+    this.#indexes = new Set()
+  }
+
+  /** @param {UnknownLink} cid */
+  async get (cid) {
+    // console.log(`index get: ${cid}`)
```
Comment (vasco-santos) [inline src/lib/car-index.js:175]:
```suggestion
```

File: src/lib/car-index.js [hunk 2]
```diff
@@ -147,3 +151,71 @@ export class StreamingCarIndex {
...
+    const key = cidToKey(cid)
+    let indexItem = this.#cache.get(key)
+    if (indexItem) {
+      // console.log(`index cache HIT: ${indexItem.origin}: ${cid} @ ${indexItem.offset}`)
```
Comment (vasco-santos) [inline src/lib/car-index.js:179]:
```suggestion
```

File: src/lib/car-index.js [hunk 3]
```diff
@@ -147,3 +151,71 @@ export class StreamingCarIndex {
...
+      if (cid.code !== raw.code) {
+        await this.#readIndex(cid)
+      }
+    } else {
+      await this.#readIndex(cid)
+      indexItem = this.#cache.get(key)
+      if (!indexItem) return // weird huh!?
```
Comment (vasco-santos) [inline src/lib/car-index.js:186]: well, if it is not found is not put in cache right?

File: src/lib/car-index.js [hunk 4]
```diff
@@ -147,3 +151,71 @@ export class StreamingCarIndex {
...
+  }
+
+  /**
+   * @param {import('multiformats').UnknownLink} cid
+   */
+  async #readIndex (cid) {
+    const key = cidToKey(cid)
+    if (this.#indexes.has(key)) return
+
+    // console.log(`reading block index: ${key}`)
```
Comment (vasco-santos) [inline src/lib/car-index.js:198]:
```suggestion
```

File: src/lib/car-index.js [hunk 5]
```diff
@@ -147,3 +151,71 @@ export class StreamingCarIndex {
...
+      if (!indexItem) return // weird huh!?
+    }
+    return { cid, ...indexItem }
+  }
+
+  /**
+   * @param {import('multiformats').UnknownLink} cid
+   */
+  async #readIndex (cid) {
+    const key = cidToKey(cid)
```
Comment (vasco-santos) [inline src/lib/car-index.js:195]: nit: could be nice to name this key as `indexesKey`. I initially thought it was `bucketKey`. But then saw it was not.

File: src/lib/car-index.js [hunk 6]
```diff
@@ -147,3 +151,71 @@ export class StreamingCarIndex {
...
+      const rawCid = Link.create(raw.code, entry.multihash)
+      // console.log(`${cid}: ${value.origin}: ${rawCid} @ ${value.offset}`)
+      this.#cache.set(cidToKey(rawCid), entry)
+    }
+    this.#indexes.add(key)
+  }
+}
+
+/** @returns {RawCIDString} */
+const cidToKey = (/** @type {import('multiformats').UnknownLink} */ cid) => {
```
Comment (vasco-santos) [inline src/lib/car-index.js:218]: nit: maybe call this `normalizeCid` or similar instead? Key makes it look like bucket keys

**Review by olizilla — APPROVED** [review]
Wild! Great work. 🚀

nits listed inline

File: src/lib/car-index.js [hunk 7]
```diff
@@ -1,11 +1,15 @@
+import * as raw from 'multiformats/codecs/raw'
+import * as Link from 'multiformats/link'
 import { base58btc } from 'multiformats/bases/base58'
 import { UniversalReader } from 'cardex/universal'
+import { MultiIndexReader } from 'cardex/multi-index'
 import defer from 'p-defer'
 
 /**
  * @typedef {import('multiformats').UnknownLink} UnknownLink
  * @typedef {import('cardex/multi-index/api').MultiIndexItem & import('cardex/multihash-index-sorted/api').MultihashIndexItem} IndexEntry
```
[the remark this answers is not in the document]
Comment (alanshaw) [reply src/lib/car-index.js:9]: Not sure if this makes it easier to understand but this is the type heirarchy:

```ts
interface IndexItem {
  digest: Uint8Array
  offset: number
}

interface MultihashIndexItem extends IndexItem {
  multihash: MultihashDigest
}

interface MultiIndexItem extends IndexItem {
  origin: CARLink
}
```

So `MultihashIndexItem`, `MultiIndexItem` are both `IndexItem`s, it's just `MultihashIndexItem` have a `multihash` property and `MultiIndexItem` have a `origin`.

I should probably change the `MultiIndexItem` type to be something like this.

```ts
type MultiIndexItem = (IndexItem & { origin: CARLink }) | (MultihashIndexItem & { origin: CARLink })
``` 

...but I'm not sure that's helping.

File: src/lib/car-index.js [hunk 8]
```diff
@@ -147,3 +151,71 @@ export class StreamingCarIndex {
...
+      this.#cache.set(cidToKey(rawCid), entry)
+    }
+    this.#indexes.add(key)
+  }
+}
+
+/** @returns {RawCIDString} */
+const cidToKey = (/** @type {import('multiformats').UnknownLink} */ cid) => {
+  if (cid.code === raw.code) return cid.toString()
+  return Link.create(raw.code, cid.multihash).toString()
```
[the remark this answers is not in the document]
Comment (alanshaw) [reply src/lib/car-index.js]: Agreed - I switched.

File: src/lib/car-index.js [hunk 9]
```diff
@@ -147,3 +151,67 @@ export class StreamingCarIndex {
...
+      if (cid.code !== raw.code) {
+        await this.#readIndex(cid)
+      }
+    } else {
+      await this.#readIndex(cid)
+      indexItem = this.#cache.get(key)
+      if (!indexItem) return
+    }
+    return { cid, ...indexItem }
+  }
```
[the remark this answers is not in the document]
Comment (alanshaw) [reply src/lib/car-index.js]: added comments, hopefully just the right amount 😀

**Pull request merged**